docs/src/cli/deploy-a-program.md

```diff
@@ -0,0 +1,217 @@
+original deployment.  Doing so leaves room for further redeployments.  But, if
+the initially deployed program is very small (like a simple helloworld program)
+and then later grows substantially, the redeployment may fail.  To avoid this,
+specify a `max_len` that is at least the size (in bytes) that the program is
+expected to become (plus some wiggle room).
+
+```bash
+solana program deploy --max-len 200000 <PROGRAM_FILEPATH>
+```
+
+Note that program accounts are required to be
+[rent-excempt](developing/programming-model/accounts.md#rent-exemption), and the
+`max-len` is fixed after initial deployment, so any SOL in the program accounts
+is locked up permanently.
+
+### Set a program's upgrade authority
+
+The program's upgrade authority must to be present to deploy a program.  If no
+authority is specified during program deployment, the default keypair is used as
+the authority.  This is why redeploying a program in the steps above didn't
+require an authority to be explicitly specified.
+
+The authority can be specified during deployment:
+
+```bash
+solana program deploy --upgrade-authority <UPGRADE_AUTHORITY_SIGNER> <PROGRAM_FILEPATH>
+```
+
+Or after deployment and using the default keypair as the current authority:
+
+```bash
+solana program set-upgrade-authority <PROGRAM_ADDRESS> --new-upgrade-authority <NEW_UPGRADE_AUTHORITY>
+```
+
+Or after deployment and specifying the current authority:
+
+```bash
+solana program set-upgrade-authority <PROGRAM_ADDRESS> --upgrade-authority <UPGRADE_AUTHORITY_SIGNER> --new-upgrade-authority <NEW_UPGRADE_AUTHORITY>
+```
+
+### Immutable programs
+
+A program can be marked immutable, which prevents all further redeployments by
+specifying the `--final` flag during deployment:
+
+```bash
+solana program deploy <PROGRAM_FILEPATH> --final
+```
+
+Or anytime after:
+
+```bash
+solana program set-upgrade-authority <PROGRAM_ADDRESS> --final
+```
+
+### Dumping a program to a file
+
+The deployed program may be dumped back to a local file:
+
+```bash
+solana program dump <ACCOUNT_ADDRESS> <OUTPUT_FILEPATH>
+```
+
+The dumped file will be in the same as what was deployed, so in the case of a
+shared object, the dumped file will be a fully functional shared object.  Note
+that the `dump` command dumps the entire data space, which means the output file
+will have trailing zeros after the shared object's data up to `max_len`.
+Sometimes it is useful to dump and compare a program to ensure it matches a
+known program binary.  The original program file can be zero-extended, hashed,
+and compared to the hash of the dumped file.
+
+```bash
+$ solana dump <ACCOUNT_ADDRESS> dump.so
+$ cp original.so extended.so
+$ truncate -r dump.so extended.so
+$ sha256sum extended.so dump.so
+```
+
+### Using an intermediary Buffer account
+
+Instead of deploying directly to the program account, the program can be written
+to an intermediary buffer account.  Doing so can be useful for things like
+multi-entity governed programs.
```

Comment:
I still don't really understand when and why I would use a buffer account. I assume it's something about separating the write to the chain (and possible updates) from the actual activation. Perhaps just half a sentence to expand on how deployment of a multi-entity governed program works?